local-postgresql-copy/README.md

```diff
@@ -0,0 +1,309 @@
+| `copy-data.sh`             | Copies all data from a Cloud Spanner database to a PostgreSQL database. This script can be used to create a copy of a Cloud Spanner database.                                                                                                                                                                                                                                |
+| `import-foreign-schema.sh` | Imports the schema of a Cloud Spanner database as a foreign schema in a PostgreSQL database. This can be used to query Cloud Spanner database directly in a PostgreSQL database. This allows the use of specific PostgreSQL features with Cloud Spanner data, such as user-defined functions, stored procedures or system functions that are not available in Cloud Spanner. |
+| `run-pg-dump.sh`           | Copies all data from the Cloud Spanner database to the local PostgreSQL database, exports the schema of the Cloud Spanner database to a `schema.sql` file, and runs `pg_dump` on the local PostgreSQL database. This can be used to reconstruct the Cloud Spanner database at a later moment.                                                                                |
+| `run-pg-restore.sh`        | Runs `pg_restore` on the local PostgreSQL database using a previously generated dump, executes the `schema.sql` file on the Cloud Spanner database to re-create the schema, and copies all data from the local PostgreSQL database to Cloud Spanner.                                                                                                                         |
+
+These scripts can also be used to copy a Cloud Spanner database to a PostgreSQL database running on
+an existing server or to a CloudSQL instance.
+
+## Try it out
+
+Follow these steps to try it out:
+
+1. Edit the `.env` file in this directory to match your Cloud Spanner setup. These variables must be
+   set to the correct values:
+
+```
+# The full path of the Google Cloud credentials file that should be used for PGAdapter.
+GOOGLE_APPLICATION_CREDENTIALS=/local/path/to/credentials.json
+
+# Project, instance and database for Cloud Spanner.
+GOOGLE_CLOUD_PROJECT=my-project
+SPANNER_INSTANCE=my-instance
+SPANNER_DATABASE=my-database
+```
+
+2. Start the Docker containers.
+
+```shell
+docker compose --env-file .env up -d
+```
+
+3. Start an interactive `psql` shell on the local PostgreSQL database. The command below assumes
+   that the `POSTGRES_CONTAINER_NAME` was left unmodified with its default value `postgres` in the `.env` file.
+
+```shell
+docker exec -it postgres /bin/bash -c 'psql -U "${POSTGRES_USER}" -d "${POSTGRES_DB}"'
```

Comment:
Minor nit, could this be `/bin/sh` and save a few cycles?

Reply:
It could, but it would require that the user has `psql` installed on localhost. This has all been set up with the assumption that the user does not need to have any PostgreSQL tools installed locally, and only requires that Docker is available. So it is really trying to be as much 'anyone can get this running without installing anything' as possible. The only requirement is the aforementioned dependency on Docker.

There is an example for how to connect to the PostgreSQL and Cloud Spanner database directly from localhost a bit further down in the document: https://github.com/GoogleCloudPlatform/pgadapter/blob/ba1499aebb6940ee2514607beec322a29341043f/local-postgresql-copy/README.md#connecting-to-the-containers

Reply:
Why would this change require `psql` to be installed?

Just to clarify, I'm not suggesting to replace the `docker exec` with `/bin/sh`, just the `/bin/bash`.

Reply:
Done